zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java

```diff
@@ -0,0 +1,295 @@
+    this(null, null, -1L, false);
+  }
+
+  /**
+   * Constructor
+   * @param storageProvider the backup provider from which to restore the backups
+   * @param snapLog the snap and log provider to use
+   * @param zxidToRestore the zxid upto which to restore, or Long.MAX to restore to the latest
+   *                      available transactionally consistent zxid.
+   * @param dryRun whether this is a dryrun in which case no files are actually copied
+   */
+  RestoreFromBackupTool(BackupStorageProvider storageProvider, FileTxnSnapLog snapLog,
+      long zxidToRestore, boolean dryRun) {
+
+    filesToCopy = new ArrayList<BackupFileInfo>();
+    snapNeededIndex = -1;
+    mostRecentLogNeededIndex = -1;
+    oldestLogNeededIndex = -1;
+
+    this.storage = storageProvider;
+    this.zxidToRestore = zxidToRestore;
+    this.snapLog = snapLog;
+    this.dryRun = dryRun;
+  }
+
+  /**
+   * Attempts to perform a restore.
+   */
+  public void run() throws IOException {
+    if (!findFilesToRestore()) {
+      throw new RestoreException("Failed to find a valid snapshot and logs to restore.");
+    }
+
+    // Create a temporary dir in backup storage for backup file processing
+    File tempDirForRestoration =
+        new File(filesToCopy.listIterator().next().getBackedUpFile().getParent(),
+            "Restore_" + Long.toHexString(zxidToRestore));
+
+    try {
+      storage.processBackupFilesForRestoration(tempDirForRestoration, filesToCopy, zxidToRestore);
+      LOG.info("Copied " + filesToCopy.size() + " backup files to temp dir " + tempDirForRestoration
+          .getPath() + " in backup storage for file preparation for restoration to zxid " + zxidToRestore + ".");
+
+      List<File> processedFiles = BackupStorageUtil.readFilesRecursivelyInDirectory(tempDirForRestoration);
+      for (File processedFile : processedFiles) {
+        String fileName = processedFile.getName();
+        File parentFile = processedFile.getParentFile();
+        File base = Util.isSnapshotFileName(fileName) ? snapLog.getSnapDir() : snapLog.getDataDir();
+        File destination = new File(base, fileName);
+
+        if (!destination.exists()) {
+          LOG.info("Copying " + fileName + " to " + destination.getPath() + ".");
+
+          if (!dryRun) {
+            // Adding "RESTORE_" prefix to file name to differentiate the restore files from original backup files,
+            // so in storage side it will use a different way to construct the file path from original backup files
+            storage.copyToLocalStorage(
+                new File(parentFile, BackupStorageUtil.RESTORE_FILE_PREFIX + fileName),
+                destination);
+          }
+        } else {
+          LOG.info("Skipping " + fileName + " because it already exists as " + destination.getPath()
+              + ".");
+        }
+      }
+    } catch (IOException ioe) {
+      throw new IOException("Hit exception when attempting to restore." + ioe.getMessage());
```

Comment:
Do we really need to catch and rethrow here since it's the same IOException? Should we just let the original exception be thrown?